passes/opt/opt_share.cc

```diff
@@ -0,0 +1,661 @@
+
+		  if (mux->type.in("$mux", "$_MUX_")) {
+			  mux_port_size = mux->getPort("\\A").size();
+			  sig = RTLIL::SigSpec{mux->getPort("\\B"), mux->getPort("\\A")};
+		  } else {
+			  mux_port_size = mux->getPort("\\A").size();
+			  sig = mux->getPort("\\B");
+		  }
+
+		  auto mux_insig = assign_map(sig);
+
+		  for (int i = 0; i < mux_insig.size(); ++i) {
+			  if (op_aux_to_outsig.count(mux_insig[i])) {
+				  remove_outsig_from_aux_bit(mux_insig[i]);
+				  continue;
+			  }
+
+			  if (!op_outbit_to_outsig.count(mux_insig[i]))
+				  continue;
+
+			  auto op_outsig = op_outbit_to_outsig.at(mux_insig[i]);
+
+			  if (op_mux_conn_map.count(op_outsig)) {
+					remove_outsig(op_outsig);
+				  continue;
+			  }
+
+			  int mux_port_id = i / mux_port_size;
+			  int mux_port_offset = i % mux_port_size;
+
+			  int op_outsig_offset;
+			  for (op_outsig_offset = 0; op_outsig[op_outsig_offset] != mux_insig[i]; ++op_outsig_offset)
+				  ;
+
+			  int j = op_outsig_offset;
+			  do {
+				  if (!op_outbit_to_outsig.count(mux_insig[i]))
+					  break;
+
+				  if (op_outbit_to_outsig.at(mux_insig[i]) != op_outsig)
+					  break;
+
+				  ++i;
+				  ++j;
+			  } while ((i / mux_port_size == mux_port_id) && (j < op_outsig.size()));
+
+			  int op_conn_width = j - op_outsig_offset;
+			  OpMuxConn inp = {
+			    op_outsig.extract(op_outsig_offset, op_conn_width),
+			    mux,
+			    outsig_to_operator.at(op_outsig),
+			    mux_port_id,
+			    mux_port_offset,
+			    op_outsig_offset,
+			  };
+
+			  op_mux_conn_map[op_outsig] = inp;
+
+			  --i;
+		  }
+	  };
+
+	std::function<void(RTLIL::SigSpec)> remove_connected_ops = [&](RTLIL::SigSpec sig) {
+		auto mux_insig = assign_map(sig);
+		for (auto outbit : mux_insig) {
+			if (op_aux_to_outsig.count(outbit)) {
+				remove_outsig_from_aux_bit(outbit);
+				continue;
+			}
+
+			if (!op_outbit_to_outsig.count(outbit))
+				continue;
+
+			remove_outsig(op_outbit_to_outsig.at(outbit));
+		}
+	};
+
+	for (auto cell : module->cells()) {
+		if (cell->type.in("$mux", "$_MUX_", "$pmux")) {
+			remove_connected_ops(cell->getPort("\\S"));
+			find_op_mux_conns(cell);
+		} else {
+			for (auto &conn : cell->connections())
+				if (cell->input(conn.first))
+					remove_connected_ops(conn.second);
+		}
+	}
+
+	for (auto w : module->wires()) {
+		if (!w->port_output)
+			continue;
+
+		remove_connected_ops(w);
+	}
+
+	return op_mux_conn_map;
+}
+
+struct OptSharePass : public Pass {
+	OptSharePass() : Pass("opt_share", "merge mutually exclusive cells of the same type that share an input signal") {}
+	void help() YS_OVERRIDE
+	{
+		//   |---v---|---v---|---v---|---v---|---v---|---v---|---v---|---v---|---v---|---v---|
+		log("\n");
+		log("    opt_share [selection]\n");
+		log("\n");
+
+		log("This pass identifies mutually exclusive cells of the same type that:\n");
+		log("    (a) share an input signal\n");
+		log("    (b) drive the same $mux, $_MUX_, or $pmux multiplexing cell allowing\n");
```

Comment:
Sorry I wasn't clear. I was trying to suggest that "allowing the cell to be merged..." does not belong inside the list and should be on the next line with no indent.

Reply:
Hope I got it right this time :)

Reply:
I don't see a change yet?